SingularityService/src/main/java/com/hubspot/singularity/mesos/SingularitySlaveAndRackHelper.java

```diff
@@ -116,4 +118,57 @@ public boolean containsAllAttributes(Map<String, String> attributes, Map<String,
     return true;
   }
 
+  public boolean containsAtLeastOneMatchingAttribute(Map<String, String> attributes, Map<String, String> otherAttributes) {
+    for (Map.Entry<String, String> entry : otherAttributes.entrySet()) {
+      if ((attributes.containsKey(entry.getKey()) && attributes.get(entry.getKey()).equals(entry.getValue()))) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public enum CpuMemoryPreference {
+    AVERAGE,
+    HIGH_MEMORY,
+    HIGH_CPU
+  }
+
+  public CpuMemoryPreference getCpuMemoryPreferenceForRequest(RequestUtilization requestUtilization) {
+    double cpuMemoryRatioForRequest = getCpuMemoryRatioForRequest(requestUtilization);
+
+    if (cpuMemoryRatioForRequest > configuration.getHighCpuSlaveCutOff()) {
+      return CpuMemoryPreference.HIGH_CPU;
+    } else if (cpuMemoryRatioForRequest < configuration.getHighMemorySlaveCutOff()) {
+      return CpuMemoryPreference.HIGH_MEMORY;
+    }
+    return CpuMemoryPreference.AVERAGE;
+  }
+
+  public CpuMemoryPreference getCpuMemoryPreferenceForSlave(SingularityOfferHolder offerHolder) {
+    double cpuMemoryRatioForSlave = getCpuMemoryRatioForSlave(offerHolder);
+    if (cpuMemoryRatioForSlave > configuration.getHighCpuSlaveCutOff()) {
+      return CpuMemoryPreference.HIGH_CPU;
+    } else if (cpuMemoryRatioForSlave < configuration.getHighMemorySlaveCutOff()) {
+      return CpuMemoryPreference.HIGH_MEMORY;
+    }
+    return CpuMemoryPreference.AVERAGE;
+  }
+
+  private double getCpuMemoryRatioForSlave(SingularityOfferHolder offerHolder) {
+    double memory = MesosUtils.getMemory(offerHolder.getCurrentResources(), Optional.empty());
+    double cpus = MesosUtils.getNumCpus(offerHolder.getCurrentResources(), Optional.empty());
+    return memory/cpus;
+  }
+
+  private double getCpuMemoryRatioForRequest(RequestUtilization requestUtilization) {
+    double cpuUsageForRequest = getEstimatedCpuUsageForRequest(requestUtilization);
+    double memUsageForRequest = requestUtilization.getAvgMemBytesUsed();
+    return cpuUsageForRequest/memUsageForRequest;
+  }
+
+  private double getEstimatedCpuUsageForRequest(RequestUtilization requestUtilization) {
+    // To account for cpu bursts, tend towards max usage if the app is consistently over-utilizing cpu, tend towards avg if it is over-utilized in short bursts
```

Comment:
if this was copied from the other offers class, can we make sure we don't duplicate it? Maybe give it a new home in another class